src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -583,12 +588,13 @@ InstanceKlass* SystemDictionary::handle_parallel_loading(JavaThread* current,
         // the original thread completes the class loading or fails
         // If it completes we will use the resulting InstanceKlass
         // which we will find below in the systemDictionary.
-        oldprobe = NULL;  // Other thread could delete this placeholder entry
 
         if (lockObject.is_null()) {
           SystemDictionary_lock->wait();
-        } else {
+        } else if (EnableWaitForParallelLoad) {
           double_lock_wait(current, lockObject);
+        } else {
+          return NULL;
```

Comment:
Not clear why we return NULL here rather than just falling through and retrying?

Reply:
If we don't return we'll actually do a wait and will hang.

Reply:
Okay but isn't that what should be expected? This is a flag to prevent a certain kind of classloader from deadlocking, so if we have that kind of classloader and the flag is disabled, then we should expect deadlock. Otherwise what will the failure mode be here? If we are introducing a new failure mode then it needs to be documented in the CSR request and possibly a Release Note.

Reply:
Coleen has reminded me that I am confused about what the workaround was doing - this code doesn't prevent deadlock, it prevents a LinkageError that would otherwise occur because the classloader has done something tricky (released the monitor) to prevent deadlock. So the correct behaviour here is proceed with the class loading and trigger the LinkageError in the other thread.

Reply:
No, the expectation without this code is a LinkageError: dupicate class definition.  I just started the release-note subtask to describe this.  The expected behavior is that we don't wait, we just let the application load the class even though they've broken the lock.

Reply:
See release note in progress: https://bugs.openjdk.org/browse/JDK-8296446